### PR TITLE
ruby/python: Shared module for typetracking through flow summaries

### PR DESCRIPTION
**Pull request by yoff**
Shared module provides predicates for
- levelStepNoCall
- basicLoadStep
- basicStoreStep
- basicLoadStoreStep
- basicWithoutContentStep
- basicWithContentStep
 
see [the output signature](https://github.com/github/codeql/pull/13178/files#diff-415d7ad85ba3d1a8d32985ee294279dbf33e0404ddac6adb038d3318aa25b4d5R78-R107).

It is shared via `identicalFiles.json` and used in ruby and python.

Will require performance testing to ensure we did not break ruby performance.

TODO:
- We will not get full benefit of this in Python until type tracking uses dataflow content (rather than just strings representing attributes, as it does now).

### CODE REVIEW

**Comment by yoff** [comment]
Force-pushed to resolve conflict. (Also added change note.)

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll [hunk 1]
```diff
@@ -0,0 +1,382 @@
+/**
+ * Provides the implementation of a summary type tracker, that is type tracking through flow summaries.
+ * To use this, you must implement the `Input` signature. You can then use the predicates in the `Output`
+ * signature to implement the predicates of the same names inside `TypeTrackerSpecific.qll`.
+ */
+
+/** The classes and predicates needed to generate a summary type tracker. */
```
Comment (asgerf) [inline ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll:7]:
```suggestion
/** The classes and predicates needed to generate type-tracking steps from summaries. */
```

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll [hunk 2]
```diff
@@ -0,0 +1,382 @@
+/**
+ * Provides the implementation of a summary type tracker, that is type tracking through flow summaries.
```
Comment (asgerf) [inline ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll:2]:
```suggestion
 * Provides the implementation of type tracking steps through flow summaries.
```
There is no such thing as "a summary type tracker"

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+  /** Gets a dataflow node respresenting the parameter of `callable` indicated by `param`. */
+  Node parameterOf(Node callable, SummaryComponent param);
+
+  /** Gets a dataflow node respresenting the return of `callable` indicated by `return`. */
+  Node returnOf(Node callable, SummaryComponent return);
+
+  // Specific summary handling
+  /** Holds if component should be treated as a level step by type tracking. */
+  predicate componentLevelStep(SummaryComponent component);
```
Comment (asgerf) [inline ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll:85]: This should be moved to the implementation, not in the input signature.

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+
+  /** Holds if the given component can't be evaluated by `evaluateSummaryComponentStackLocal`. */
+  predicate isNonLocal(SummaryComponent component);
```
Comment (asgerf) [inline ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll:88]: Same, should be moved to the implementation, not in the input signature.

**Review by asgerf — COMMENTED** [review]
Thanks for doing this! The signature does end up being rather verbose but I guess that's just the way it is. A few minor things otherwise LGTM.

**Comment by yoff** [comment]
Thanks for the review, indeed it looks like we can remove those two predicates from the interface. Nice! 💪

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll (hunk 3 above)
Comment (yoff) [reply ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll]: Done

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll (hunk 4 above)
Comment (yoff) [reply ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll]: Done

**Comment by yoff** [comment]
The merge includes the results of https://github.com/github/codeql/pull/12964/files#diff-9547dcadaf57e980b53a87f8f1f5124c65ca46cfe537c69cda27d142078663c1

**Comment by yoff** [comment]
So _now_ this PR is ready for review again :-)

File: python/ql/test/experimental/dataflow/typetracking-summaries/TestSummaries.qll [hunk 5]
```diff
@@ -0,0 +1,189 @@
+private import python
+private import semmle.python.dataflow.new.FlowSummary
+private import semmle.python.ApiGraphs
+
+/**
+ * This module ensures that the `callStep` predicate in
+ * our type tracker implelemtation does not refer to the
```
Comment (asgerf) [inline python/ql/test/experimental/dataflow/typetracking-summaries/TestSummaries.qll:7]:
```suggestion
 * our type tracker implementation does not refer to the
```

File: python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll [hunk 6]
```diff
@@ -144,3 +163,99 @@ predicate basicWithContentStep(Node nodeFrom, Node nodeTo, ContentFilter filter)
...
+  DataFlowPublic::CallCfgNode call, DataFlowPublic::Node arg,
+  DataFlowDispatch::ParameterPosition ppos
+) {
+  exists(DataFlowDispatch::ArgumentPosition apos |
+    DataFlowDispatch::parameterMatch(ppos, apos) and
+    DataFlowDispatch::normalCallArg(call.getNode(), arg, apos)
+  )
+}
+
+module SummaryTypeTrackerInput implements SummaryTypeTracker::Input {
```
Comment (asgerf) [inline python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll:182]: Shouldn't this be private?

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll [hunk 7]
```diff
@@ -514,101 +347,105 @@ predicate isNonLocal(SummaryComponent component) {
...
-    or
-    callable.propagatesFlow(_, stack, true)
-    or
-    // include store summaries as they may skip an initial step at the input
-    hasStoreSummary(callable, _, stack, _)
-  )
-  or
-  dependsOnSummaryComponentStackCons(callable, _, stack)
-}
+module SummaryTypeTrackerInput implements SummaryTypeTracker::Input {
```
Comment (asgerf) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:353]: Should be private?

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll [hunk 8]
```diff
@@ -514,101 +347,105 @@ predicate isNonLocal(SummaryComponent component) {
...
-    )
-  )
+      callable.asExpr().getExpr()
+  }
+
+  // Relating callables to nodes
+  Node callTo(SummarizedCallable callable) { result.asExpr().getExpr() = callable.getACallSimple() }
 }
+
+module TypeTrackerSummaryFlow = SummaryTypeTracker::SummaryFlow<SummaryTypeTrackerInput>;
```
Comment (asgerf) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:451]: Also private?

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll [hunk 9]
```diff
@@ -0,0 +1,389 @@
+/**
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll:1]: I think this file should be in an `internal` folder.

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+  /** Gets a dataflow node respresenting a call to `callable`. */
+  Node callTo(SummarizedCallable callable);
+}
+
+/**
+ * The predicates provided by a summary type tracker.
+ * These are meant to be used in `TypeTrackerSpecific.qll`
+ * inside the predicates of the same names.
+ */
+signature module Output<Input I> {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll:93]: I don't think there is any need to provide an explicit output signature, when the output is not used itself as input to another parameterized module.

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+ * Implementation of the summary type tracker, that is type tracking through flow summaries.
+ */
+module SummaryFlow<Input I> implements Output<I> {
+  pragma[nomagic]
+  private predicate hasLoadSummary(
+    I::SummarizedCallable callable, I::TypeTrackerContent contents, I::SummaryComponentStack input,
+    I::SummaryComponentStack output
+  ) {
+    callable.propagatesFlow(I::push(I::content(contents), input), output, true) and
+    not isNonLocal(input.head()) and
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll:142]: I prefer to have predicate defined in the file before they are used (mutually recursive predicates being, of course, an exception). So I would move `isNonLocal` to the beginning.

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+
+  pragma[nomagic]
+  private predicate isNonLocal(I::SummaryComponent component) {
+    component = I::content(_)
+    or
+    component = I::withContent(_)
+  }
+
+  /**
+   * Gets a data flow I::Node corresponding an argument or return value of `call`,
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll:221]: Missing backticks.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll [hunk 13]
```diff
@@ -514,101 +347,105 @@ predicate isNonLocal(SummaryComponent component) {
...
+    ) and
+    result = MkElementFilter()
+  }
+
+  // Summaries and their stacks
+  class SummaryComponent = FlowSummary::SummaryComponent;
+
+  class SummaryComponentStack = FlowSummary::SummaryComponentStack;
+
+  SummaryComponentStack singleton(SummaryComponent component) {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:397]: Couldn't you just have used a predicate alias here?
`predicate SummaryComponentStack = FlowSummary::SummaryComponentStack::singleton/1`.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll [hunk 14]
```diff
@@ -514,101 +347,105 @@ predicate isNonLocal(SummaryComponent component) {
...
+    result = FlowSummary::SummaryComponentStack::singleton(component)
+  }
+
+  SummaryComponentStack push(SummaryComponent component, SummaryComponentStack stack) {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:401]: Same about alias.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll [hunk 15]
```diff
@@ -514,101 +347,105 @@ predicate isNonLocal(SummaryComponent component) {
...
+    result = FlowSummary::SummaryComponentStack::push(component, stack)
+  }
+
+  // Relating content to summaries
+  SummaryComponent content(TypeTrackerContent contents) {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:406]: Same about alias.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll [hunk 16]
```diff
@@ -514,101 +347,105 @@ predicate isNonLocal(SummaryComponent component) {
...
+    result = FlowSummary::SummaryComponent::content(contents)
+  }
+
+  SummaryComponent withoutContent(TypeTrackerContent contents) {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:410]: Same about alias.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll [hunk 17]
```diff
@@ -514,101 +347,105 @@ predicate isNonLocal(SummaryComponent component) {
...
+    result = FlowSummary::SummaryComponent::withoutContent(contents)
+  }
+
+  SummaryComponent withContent(TypeTrackerContent contents) {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:414]: Same about alias.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll [hunk 18]
```diff
@@ -514,101 +347,105 @@ predicate isNonLocal(SummaryComponent component) {
...
+    result = FlowSummary::SummaryComponent::withContent(contents)
+  }
+
+  SummaryComponent return() { result = FlowSummary::SummaryComponent::return() }
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:418]: Same about alias.

**Review by hvitved — COMMENTED** [review]
Looks great, thanks for making this code shared. I have a few trivial comments. We should also do final DCA runs before merging.

File: python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll (hunk 6 above)
Comment (yoff) [reply python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll:182]: Ah, yes, that would be nicer.

File: python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll [hunk 19]
```diff
@@ -144,3 +163,99 @@ predicate basicWithContentStep(Node nodeFrom, Node nodeTo, ContentFilter filter)
...
+    // result should be return value of callable which should be a lambda
+    result.asCfgNode() =
+      callable.getALocalSource().asExpr().(CallableExpr).getInnerScope().getAReturnValueFlowNode()
+  }
+
+  // Relating callables to nodes
+  Node callTo(SummarizedCallable callable) { result = callable.getACallSimple() }
+}
+
+module TypeTrackerSummaryFlow = SummaryTypeTracker::SummaryFlow<SummaryTypeTrackerInput>;
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll:261]: Actually, yes, the outside does not need this, I think.

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll (hunk 7 above)
Comment (yoff) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:353]: yes

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll (hunk 8 above)
Comment (yoff) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:451]: yes

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll (hunk 13 above)
Comment (yoff) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll:397]: I remember having some issues with using an alias, but I may have just been fumbling. I will try again :-)

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll (hunk 10 above)
Comment (yoff) [reply ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll:93]: Might it be in the future, though? If we get type tracking as a parameterised module (as started by Java).

File: ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll (hunk 13 above)
Comment (yoff) [reply ruby/ql/lib/codeql/ruby/typetracking/TypeTrackerSpecific.qll]: That worked fine...and looks a lot neater..

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll (hunk 10 above)
Comment (yoff) [reply ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll]: I mean, it is easy enough to add when needed (so we could take it out now), but I actually like how it codifies the intention of the module...

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll (hunk 11 above)
Comment (yoff) [reply ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll]: done

File: ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll (hunk 12 above)
Comment (yoff) [reply ruby/ql/lib/codeql/ruby/typetracking/SummaryTypeTracker.qll]: done

**Review by hvitved — DISMISSED** [review]
LGTM. Let's do a final DCA run before merging.

**Comment by yoff** [comment]
> LGTM. Let's do a final DCA run before merging.

Agreed, I have kept the `Awaiting evaluation` label. See running experiments in backlinks.

File: python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll [hunk 20]
```diff
@@ -144,3 +163,99 @@ predicate basicWithContentStep(Node nodeFrom, Node nodeTo, ContentFilter filter)
...
+        exists(int i | ppos.isPositional(i) |
+          p = callable.getALocalSource().asExpr().(CallableExpr).getInnerScope().getArg(i)
+        )
+        or
+        exists(string name | ppos.isKeyword(name) |
+          p = callable.getALocalSource().asExpr().(CallableExpr).getInnerScope().getArgByName(name)
+        )
+      )
+    )
+  }
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll:248]: Let's talk about these two offline. I would like to understand the interaction with the call-graph a bit better :blush:

File: python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll [hunk 21]
```diff
@@ -144,3 +163,99 @@ predicate basicWithContentStep(Node nodeFrom, Node nodeTo, ContentFilter filter)
...
+
+  Node returnOf(Node callable, SummaryComponent return) {
+    return = FlowSummary::SummaryComponent::return() and
+    // result should be return value of callable which should be a lambda
+    result.asCfgNode() =
+      callable.getALocalSource().asExpr().(CallableExpr).getInnerScope().getAReturnValueFlowNode()
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll:254]: If this is only a Lambda, let's make the cast more explicit

```suggestion
    // result should be return value of callable which should be a lambda
    result.asCfgNode() =
      callable.getALocalSource().asExpr().(Lambda).getInnerScope().getAReturnValueFlowNode()
```

File: python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll [hunk 22]
```diff
@@ -144,3 +163,99 @@ predicate basicWithContentStep(Node nodeFrom, Node nodeTo, ContentFilter filter)
...
+  // Relating content to summaries
+  SummaryComponent content(TypeTrackerContent contents) {
+    result = FlowSummary::SummaryComponent::content(contents)
+  }
+
+  SummaryComponent withoutContent(TypeTrackerContent contents) { none() }
+
+  SummaryComponent withContent(TypeTrackerContent contents) { none() }
+
+  SummaryComponent return() { result = FlowSummary::SummaryComponent::return() }
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll:220]: Would like the nice aliases for Python as well 

```suggestion
  predicate singleton = FlowSummary::SummaryComponentStack::singleton/1;

  predicate push = FlowSummary::SummaryComponentStack::push/2;

  // Relating content to summaries
  predicate content = FlowSummary::SummaryComponent::content/1;

  predicate withoutContent = FlowSummary::SummaryComponent::withoutContent/1;

  predicate withContent = FlowSummary::SummaryComponent::withContent/1;

  predicate return = FlowSummary::SummaryComponent::return/0;
```

**Review by RasmusWL — CHANGES_REQUESTED** [review]

File: python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll (hunk 21 above)
Comment (yoff) [reply python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackerSpecific.qll:254]: I used this comment as a note. It is great to also include `FunctionExpr`, I will change the comment.

File: python/ql/test/experimental/dataflow/typetracking-summaries/TestSummaries.qll [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+    preservesValue = true
+    or
+    input = "Argument[0].ReturnValue" and
+    output = "ReturnValue" and
+    preservesValue = true
+  }
+}
+
+private class SummarizedCallableReversed extends SummarizedCallable {
+  SummarizedCallableReversed() { this = "reversed" }
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/test/experimental/dataflow/typetracking-summaries/TestSummaries.qll]: thanks for fixing :+1:

**Review by hvitved — APPROVED** [review]

**Pull request merged**